### ISSUE TITLE
DefaultExecuteContext ThreadLocals don't work correctly when nesting executions in a single thread

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#12076]
The `ThreadLocal` values in `DefaultExecuteContext` don't work correctly when nesting executions in a single thread, i.e. when keeping an open `Cursor` and running new queries before closing that open `Cursor`. An example of this is:

```java
List<Tuple2<A, B>> list =
create().selectFrom(TAuthor())
        .orderBy(TAuthor_ID())
        .stream()
        .flatMap(a -> create()
            .selectFrom(TBook())
            .where(TBook_AUTHOR_ID().eq(a.getValue(TAuthor_ID())))
            .orderBy(TBook_ID())
            .stream()
            .map(b -> Tuple.<A, B>tuple(a, b))
        )
        .collect(Collectors.toList());
```

A side effect of this is that we can't deserialise nested UDTs in Oracle. An exception like this may arise:

```
org.jooq.exception.DataAccessException: SQL [select "TEST"."T_AUTHOR"."ID", "TEST"."T_AUTHOR"."FIRST_NAME", "TEST"."T_AUTHOR"."LAST_NAME", "TEST"."T_AUTHOR"."DATE_OF_BIRTH", "TEST"."T_AUTHOR"."YEAR_OF_BIRTH", "TEST"."T_AUTHOR"."ADDRESS" from "TEST"."T_AUTHOR" order by "TEST"."T_AUTHOR"."ID" -- SQL rendered with a free trial version of jOOQ 3.15.0-SNAPSHOT]; Error while reading field: "TEST"."T_AUTHOR"."ADDRESS", at JDBC index: 6
	at org.jooq_3.15.0-SNAPSHOT.ORACLE18C.debug(Unknown Source)
	at org.jooq.impl.Tools.translate(Tools.java:2913)
	at org.jooq.impl.DefaultExecuteContext.sqlException(DefaultExecuteContext.java:643)
	at org.jooq.impl.CursorImpl$CursorIterator.fetchNext(CursorImpl.java:1439)
	at org.jooq.impl.CursorImpl$CursorIterator.hasNext(CursorImpl.java:1400)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:132)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at org.jooq.test.all.testcases.FetchTests.testResultQueryStream(FetchTests.java:1887)
	at org.jooq.test.jOOQAbstractTest.testResultQueryStream(jOOQAbstractTest.java:2014)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:93)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:529)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:756)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:452)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
Caused by: java.sql.SQLException: Error while reading field: "TEST"."T_AUTHOR"."ADDRESS", at JDBC index: 6
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.setValue(CursorImpl.java:1561)
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.apply(CursorImpl.java:1494)
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.apply(CursorImpl.java:1)
	at org.jooq.impl.RecordDelegate.operate(RecordDelegate.java:143)
	at org.jooq.impl.CursorImpl$CursorIterator.fetchNext(CursorImpl.java:1424)
	... 41 more
Caused by: java.lang.NullPointerException: Cannot invoke "org.jooq.Scope.configuration()" because "scope" is null
	at org.jooq.impl.AbstractQualifiedRecord.readSQL(AbstractQualifiedRecord.java:126)
	at oracle.sql.STRUCT.toClass(STRUCT.java:630)
	at oracle.sql.STRUCT.toJdbc(STRUCT.java:581)
	at oracle.jdbc.driver.NamedTypeAccessor.getObject(NamedTypeAccessor.java:207)
	at oracle.jdbc.driver.GeneratedStatement.getObject(GeneratedStatement.java:202)
	at oracle.jdbc.driver.GeneratedScrollableResultSet.getObject(GeneratedScrollableResultSet.java:350)
	at org.jooq.tools.jdbc.DefaultResultSet.getObject(DefaultResultSet.java:678)
	at org.jooq.tools.jdbc.DefaultResultSet.getObject(DefaultResultSet.java:678)
	at org.jooq.tools.jdbc.DefaultResultSet.getObject(DefaultResultSet.java:678)
	at org.jooq.impl.CursorImpl$CursorResultSet.getObject(CursorImpl.java:647)
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding.get0(DefaultBinding.java:3565)
	at org.jooq.impl.DefaultBinding$DefaultRecordBinding.get0(DefaultBinding.java:1)
	at org.jooq.impl.DefaultBinding$AbstractBinding.get(DefaultBinding.java:934)
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.setValue(CursorImpl.java:1551)
	... 45 more
```

The `ThreadLocals` are hadly ever used outside of Oracle's UDTs, so this is probably a rather esoteric and rare bug.

**Comment by lukaseder** [comment]
A fix will be quite delicate. Currently, we can't really have any resource leaks in this area, because every thread can only reference at most one JDBC `Connection` for the current context. If this is fixed with a `ThreadLocal<Deque<T>>` to act as a stack of resources, we might have stacks that are never cleaned.

**Comment by lukaseder** [comment]
Duplicate of https://github.com/jOOQ/jOOQ/issues/12527, fixed in jOOQ 3.16.0
